src/libopensc/sc.c

```diff
@@ -850,6 +850,17 @@ int _sc_parse_atr(sc_reader_t *reader)
 		sc_log(reader->ctx, "invalid sync byte in ATR: 0x%02X\n", p[0]);
 		return SC_ERROR_INTERNAL;
 	}
+
+	/* see if contactless */
+	if (atr_len >= 4
+			&& p[0] == 0x3b
+			&& (p[1] & 0xF0) == 0x80
+			&& p[2] == 0x80
+			&& p[3] == 0x01) {
+		reader->atr_info.contactless = 1;
```

Comment:
I think it is not correct to speak of a contact-less card in this case. The ATR is technically a contact-based propberty and PC/SC describes how to map the contact-less connection parameters to an ATR (which seems to be what you're testing here). However, I think, it is still possible that a contact-based card has this very ATR.

For avoiding this problem, we're trying to read the card's UID in reader-pcsc.c and you could take advantage of the already read value to detect a contact-less card. The UID is a contact-less property only. The drawback, however is that the command for getting the UID is not supported by all readers.

Reply:
https://github.com/OpenSC/OpenSC/pull/2053#discussion_r1102516064 is the same as https://github.com/OpenSC/OpenSC/pull/2053#discussion_r1102512477 and fixed by [PIV Use piv_free_sm_apdu to cleanup if piv_encode_apdu fails](https://github.com/OpenSC/OpenSC/pull/2053/commits/24a2001261c3a50cb05a05ad188a006c99588d14)

https://github.com/OpenSC/OpenSC/pull/2053#discussion_r1102625542 is fixed by [PIV Improve testing of AuthCryptogram](https://github.com/OpenSC/OpenSC/pull/2053/commits/1867019c95adec1df553cff3e3caa9223be507bf)

https://github.com/OpenSC/OpenSC/pull/2053#discussion_r1102630141 is fixed by [Add contactless flag to atr_info in sc_reader_t](https://github.com/OpenSC/OpenSC/pull/2053/commits/e1c97c84a9cdf37c109195ba0d27e7135503c436)
and pointed out in a discussion in:  https://github.com/frankmorgner/vsmartcard/pull/251 that the IFD makes up the ATR for contactless cards. And looking at all the ATRs listed in OpenSC, PCSC is returning what the IFD says is the ATR. 
For remote readers, the ATR would be made  up via the remote reader. It is not clear if the folloing would work:

> For avoiding this problem, we're trying to read the card's UID in reader-pcsc.c and you could take advantage of the already read value to detect a contact-less card. The UID is a contact-less property only. The drawback, however is that the command for getting the UID is not supported by all readers.

Reply:
If you point me to a document that explaines why `3B 80 80 01` can only come from a contactless card, then I'm fine with it, otherwise it is just incorrect. The UID mechanism is correct, even though it doesn't work for every reader. Especially in PIV you can probe certain files and check the card's return code for a contactless specifc response - there is no need to preemptively incorporate an incorrect detection mechanism